spring-amqp/src/test/java/org/springframework/amqp/support/converter/Jackson2JsonMessageConverterTests.java

```diff
@@ -69,222 +74,266 @@ public void before(){
-	   byte[] bytes = "{\"name\" : \"foo\" }".getBytes();
-	   MessageProperties messageProperties = new MessageProperties();
-	   Message message = new Message(bytes, messageProperties);
-	   Object foo = jsonConverterWithDefaultType.fromMessage(message);
-	   assertEquals(new String(bytes), new String((byte[]) foo));
-   }
-
-   public static class Foo {
-      private String name = "foo";
-
-      public String getName() {
-         return name;
-      }
-
-      public void setName(String name) {
-         this.name = name;
-      }
-
-      @Override
-      public int hashCode() {
-         final int prime = 31;
-         int result = 1;
-         result = prime * result + ((name == null) ? 0 : name.hashCode());
-         return result;
-      }
-
-      @Override
-      public boolean equals(Object obj) {
-         if (this == obj) {
-			return true;
+	public void testDefaultTypeConfig() {
+		byte[] bytes = "{\"name\" : \"foo\" }".getBytes();
+		MessageProperties messageProperties = new MessageProperties();
+		messageProperties.setContentType("application/json");
+		Message message = new Message(bytes, messageProperties);
+		Object foo = jsonConverterWithDefaultType.fromMessage(message);
+		assertTrue(foo instanceof Foo);
+	}
+
+	@Test
+	public void testNoJsonContentType() {
+		byte[] bytes = "{\"name\" : \"foo\" }".getBytes();
+		MessageProperties messageProperties = new MessageProperties();
+		Message message = new Message(bytes, messageProperties);
+		Object foo = jsonConverterWithDefaultType.fromMessage(message);
+		assertEquals(new String(bytes), new String((byte[]) foo));
+	}
+
+	@Test
+	public void testNoTypeInfo() {
+		byte[] bytes = "{\"name\" : { \"foo\" : \"bar\" } }".getBytes();
+		MessageProperties messageProperties = new MessageProperties();
+		messageProperties.setContentType("application/json");
+		Message message = new Message(bytes, messageProperties);
+		Object foo = this.converter.fromMessage(message);
+		assertThat(foo, instanceOf(LinkedHashMap.class));
+		@SuppressWarnings("unchecked")
+		Map<String, Object> map = (Map<String, Object>) foo;
+		assertThat(map.get("name"), instanceOf(LinkedHashMap.class));
+	}
+
+	@Test
+	public void testInferredTypeInfo() {
+		byte[] bytes = "{\"name\" : \"foo\" }".getBytes();
+		MessageProperties messageProperties = new MessageProperties();
+		messageProperties.setContentType("application/json");
+		messageProperties.setInferredArgumentType(Foo.class);
+		Message message = new Message(bytes, messageProperties);
+		Object foo = this.converter.fromMessage(message);
+		assertThat(foo, instanceOf(Foo.class));
+	}
+
+	@Test
+	public void testInferredGenericTypeInfo() throws Exception {
+		byte[] bytes = "[ {\"name\" : \"foo\" } ]".getBytes();
+		MessageProperties messageProperties = new MessageProperties();
+		messageProperties.setContentType("application/json");
+		messageProperties
+				.setInferredArgumentType(Bar.class.getMethod("setFoos", List.class).getGenericParameterTypes()[0]);
```

Comment:
I haven't seen yet how you inferre the `Type` in the annotation processor, but how about this variant:
```
messageProperties.setInferredArgumentType((new ParameterizedTypeReference<List<Foo>>() {}).getType());
```
Although that produce the same result, so may even never mind...
Just for the general info then :smile: